### ISSUE TITLE
`GraphMultisetTransformer` output changes with batch size

### ISSUE THREAD

**Issue by SantiagoMorandi** [pyg-team/pytorch_geometric#6886]
### 🐛 Describe the bug

Hello, it seems that when using the Graph Multiset Transformer `torch_geometric.nn.aggr.GraphMultisetTransformer` as pooling layer, the GNN predictions vary with the batch size, a behavior that should not occur in principle.  I found this problem while working on my own project (PyG 2.0.3), so what I did was to reproduce the same error with the TUDataset and a generic GNN, using the latest PyG and Pytorch versions. When changing the pooling layer to the simpler global_add_pool, the predictions remain constant with the batch size, as expected.

``` python
import torch
import torch_geometric
import torch.nn.functional as F
from torch_geometric.nn import GCNConv, global_add_pool
from torch_geometric.nn.aggr import GraphMultisetTransformer
from torch_geometric.datasets import TUDataset
from torch_geometric.loader import DataLoader
torch_geometric.seed_everything(3)
dataset = TUDataset(root='/tmp/ENZYMES', name='ENZYMES')

class GNN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GCNConv(dataset.num_node_features, 16)
        self.pool_gmt = GraphMultisetTransformer(16, 16, 1, num_heads=1, 
                                             num_nodes=300, pool_sequences=["GMPool_I"], 
                                             pooling_ratio=0.25, layer_norm=False)
        self.lin = torch.nn.Linear(16, 1)
        self.pool_add = global_add_pool
    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = self.pool_gmt(x, data.batch)   # GMT (comment this line to see difference with add pooling)      
        # # Add pooling layer (uncomment next 3 lines to see difference with gmt)
        # x = F.relu(x)
        # x= self.lin(x)         
        # x = self.pool_add(x, data.batch)        
        return x.view(x.size(0), -1)

model = GNN().eval()
dataloader_bs_1 = DataLoader(dataset, batch_size=1, shuffle=False)
dataloader_bs_16 = DataLoader(dataset, batch_size=16, shuffle=False)
predictions_bs1, predictions_bs16 = [], []
for batch in dataloader_bs_1:  # Batch size 1
    prediction = model(batch)
    true = batch.y
    for i in range(len(prediction)):
        my_list = [prediction[i].item(), true[i].item()]
        predictions_bs1.append(my_list)
for batch in dataloader_bs_16:  # Batch size 16
    prediction = model(batch)
    true = batch.y
    for i in range(len(prediction)):
        my_list = [prediction[i].item(), true[i].item()]
        predictions_bs16.append(my_list)
# Results should not vary with batch size, but they change. With global_add_pool, the results are constant with batch size
print(predictions_bs1[:10] == predictions_bs16[:10])
```

### Environment

* PyG version: 2.2.0
* PyTorch version: 1.13.1
* OS: Ubuntu 20.04.5 LTS
* Python version: 3.9.16
* CUDA/cuDNN version: 11.6/8302
* How you installed PyTorch and PyG (`conda`, `pip`, source): conda
* Any other relevant information (*e.g.*, version of `torch-scatter`):

**Comment by EdisonLeeeee** [comment]
Looks like this was fixed in master. Can you upgrade your PyG from master and try again?

**Comment by SantiagoMorandi** [comment]
Yes, with the version from master now the outputs do not vary anymore with the batch size, thanks a lot!
I noticed the implementation changed quite significantly, for example now is not possible to define the number of output channels in order to pass from a node-wise to graph-wise representation. In the past implementation, looks like this was done by applying two linear transformations that now are not present anymore in the GMT layer. Was maybe there the error source?

**Comment by rusty1s** [comment]
Yes, we did a complete revisit of `GraphMultisetTransformer` for PyG 2.3. If you know want to map the output of GMT to a different dimension, you would need to append a `Linear` layer on top.

**Comment by SantiagoMorandi** [comment]
ok, got it. Thank you very much!

**Comment by oliverlovros** [comment]
I have a similar issue related to the same pooling layer GraphMultiSetTransformer. I was using it in PyG 2.0.3 and noticed that test accuracy varies with the batch size (graph-wise prediction always worse than batch-wise prediction).

The new implementation in PyG 2.3.0 fixed the issue, however now the performance of my GNN worsened substantially.

My main question right now is: In the PyG 2.0.3, what was the error that caused the performance to vary with the batch size?

Thanks in advance,

Oliver

**Comment by rusty1s** [comment]
Please note that in PyG 2.3.0, you need to apply your GNN execution outside `GraphMultisetTransformer`, which might explain the difference in performance. Overall, I think we fixed the implementation and made it much more cleaner.

I can't give you a pointer right now about graph-wise vs batch-wise prediction, but you should be able to debug which operation causes the difference in outputs by comparing intermediate outputs of the two. Let me know if you need any help with that.
